File: abc055_c/Main.java
import java.io.IOException;
import java.util.Arrays;
import java.util.Scanner;

class Main {
	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		long N = sc.nextLong();
		long M = sc.nextLong();

		if (N > M / 2) {
			System.out.println(Math.min(N, M / 2));
		} else {
			long 残り = M - N * 2;
			System.out.println(N + 残り / 4);
		}
	}

	private static boolean isDebug = System.getProperty("sun.desktop") != null;

	private static void debug(Object... o) {
		if (isDebug) {
			System.err.println(Arrays.deepToString(o));
		}
	}
}

File: abc055_c/Main.kt
import java.util.*

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    val N = sc.nextLong()
    val M = sc.nextLong()

    if (N > M / 2) {
        println(Math.min(N, M / 2))
    } else {
        println(N + (M - N * 2) / 4)
    }
}
